### ISSUE TITLE
UDS connection doesn't repair after agent restart

### ISSUE THREAD

**Issue by danopia** [brightcove/hot-shots#114]
I've been testing UDS for Kubernetes applications and have noticed that restarting the agent causes the metrics calls to throw. The library doesn't seem to repair - my code is just throwing repeatedly until I restart the nodejs application.

The first call after agent restart looked like this:

```
Error: Error sending hot-shots message: Error: send -111
Error: Error sending hot-shots message: Error: Error sending hot-shots message: Error: send -111
```

After that, the rest of the attempts look like this:

```
Error: Error sending hot-shots message: Error: send -107
Error: Error sending hot-shots message: Error: Error sending hot-shots message: Error: send -107
```

Ideally these errors can be handled invisibly by reconnecting, not sure what the best course of action is here. [Datadog docs say](https://docs.datadoghq.com/developers/dogstatsd/unix_socket/#how-it-works) "When the Agent restarts, the existing socket is deleted and replaced by a new one. Client libraries detect this change and connect seamlessly to the new socket."
I used `.event()` to report JS errors, so that failed too, giving me a stacktrace:

```
(node:1) UnhandledPromiseRejectionWarning: Error: Error sending hot-shots message: Error: Error sending hot-shots message: Error: send -107
    at handleCallback (/src/node_modules/hot-shots/lib/statsd.js:369:32)
    at Client.sendMessage (/src/node_modules/hot-shots/lib/statsd.js:397:5)
    at Client._send (/src/node_modules/hot-shots/lib/statsd.js:308:12)
    at Client.send (/src/node_modules/hot-shots/lib/statsd.js:285:8)
    at Client.event (/src/node_modules/hot-shots/lib/statsFunctions.js:308:10)
    at Timeout.loop [as _onTimeout] (/src/app.js:69:9)
    at processTicksAndRejections (internal/process/task_queues.js:89:5)
```

**Comment by danopia** [comment]
After reading through https://github.com/DataDog/datadog-agent/wiki/Unix-Domain-Sockets-support#implementing-uds-support-for-statsd-client-libraries I see some key points:

* the library should act like UDP by default, aka, it should not throw or block for more than 1ms when emitting metrics to UDS
* unless a 'strict' config is enabled, dropping metrics during a downed agent connection is fine
* the library should try reconnecting regularly (but not too often) after getting disconnected

Since this library isn't directly owned by Datadog, I'm not sure how motivated you are to follow Datadog's spec...

**Comment by bdeitte** [comment]
Hey @danopia, thanks for all the details!  So I am the main maintainer of this project, and I do hope this issue can be fixed.  It does sound like an ugly one, and it'll actually be one I run into myself at Maxwell Health (where I work now) as we're looking to eventually switch over to UDS and already use this library/Kubernetes.

That being said, I don't think I'm going to be able to dig into it right away.  This project does tend to be a split between myself doing things and community support- it's about 50/50 right now.  So if you or anybody else want this fixed right away, I'm happy to provide pointers on how to write a test or anything else here.  Otherwise it will be on my list to poke at later.

**Comment by danopia** [comment]
Hey, good to hear from you quickly :) I've only configured one low-importance nodejs process for UDS so far, so for me, it's acceptable that hot-shots throws an error up the stack. This also means I have a bit of a testing ground for reproing the issue :)

I see that there's already a test around error handling. I'd be willing to try writing one, I'm not sure what aspects of closed connections are already handled, or if reconnection is implemented at all, so wouldn't mind seeing if I can easily repro the exact issue in a test. Should I configure an error handler in my application as it stands? I didn't read enough docs to see that option the other day :P

For some wider context -- I've been opening UDS issues all over the place. Turns out [the official C# library completely lacks Unix sockets](https://github.com/DataDog/dogstatsd-csharp-client/issues/85) (to no one's surprise). So I've worked out [transiting UDP traffic back to the agent over a second UDS](https://github.com/DataDog/datadog-agent/issues/3870). Adding a proxy seems like the current best bet until UDS support is widespread.

**Comment by bdeitte** [comment]
Hey @danopia well all help is definitely appreciated!  The UDS support in this library is pretty new, just so you know.  It's from https://github.com/brightcove/hot-shots/pull/107 and https://github.com/brightcove/hot-shots/pull/112

As for tests, we have tests for connections being broken and error handling correctly: https://github.com/brightcove/hot-shots/blob/master/test/errorHandling.js#L148
But nothing about a connection being broken in between connections, and it working on the second try.  I think you could fit it in there, and fairly easily get a test running that shows the issue.

In the test written, you'll need to shut down the server created I assume (this one https://github.com/brightcove/hot-shots/blob/master/test/helpers/helpers.js#L127) and you can add a new method to the helpers file that is called to do that.

To test quickly, in addition to the usual mocha "only" trick, you can also alter https://github.com/brightcove/hot-shots/blob/master/test/errorHandling.js#L148 temporarily while testing to only test one UDS variable and not every possible attempt.  And if the test is specific to UDS, that's ok- you can see in the errorHandling file how we tend to handle that.

As for what is broken, there is no attempts to reconnect.  Assuming the unix-dram library tells you when things are disconnected, you should be able to add some kind of reconnection attempt to the socket that is set up at https://github.com/brightcove/hot-shots/blob/master/lib/statsd.js#L70.  There may be some kinds of reconnection attempts in the underlying unix-dram library, but I really don't know much about that library.

Let me know if you have any more questions if you do dig into this!

**Comment by NinjaBanjo** [comment]
I took a stab at this. This is causing issues for us and would like to see it resolve.

https://github.com/brightcove/hot-shots/pull/116

**Comment by Enrico2** [comment]
@bdeitte Any chance anyone else is taking a stab at this? @NinjaBanjo mentioned he's no longer able to. If I felt more comfortable with Node, I'd jump in, but given the time it would take to go through the learning curve, I can't justify it against my own priorities. Any help would be appreciated.

**Comment by bdeitte** [comment]
@Enrico2 Unfortunately I don't know anybody else taking a stab at this right now.  I do pick up on some of the most needed items myself from time-to-time, but I can't at the moment.  It's certainly at the top of my list if I can myself.  And as always, if someone else does pick this up, I am prompt to review and ensure PRs go through quickly when done.

**Comment by msiebuhr** [comment]
I've given the changes in #123 ... and it doesn't work.

First off, `send()`'s `handleCallback()` swallows the error if a callback is set ([code](https://github.com/brightcove/hot-shots/blob/e45e0f2bb0fd917840b6ac8ee724e46632fcb5d8/lib/statsd.js#L361-L371)).

Secondly, I tried having it calling `this.socket.emit('error', err)`, but that didn't make it to the error handler for some reason or other.

Next came the error codes, that should be negative (see #128).

I ended up completely ignoring the error code and calling `this.socket.close()` whenever an error was returned in `handleCallback` and then trying to re-connect of the socket was missing in the beginning of `send()`... And now it works for me. Patchy and noisy, but working.
